mithril-aggregator/src/services/certifier/buffered_certifier.rs

```diff
@@ -0,0 +1,654 @@
+    multi_signer: Arc<dyn MultiSigner>,
+    buffered_single_signature_store: Arc<dyn BufferedSingleSignatureStore>,
+    logger: Logger,
+}
+
+impl BufferedCertifierService {
+    /// Create a new instance of `BufferedCertifierService`.
+    pub fn new(
+        certifier_service: Arc<dyn CertifierService>,
+        multi_signer: Arc<dyn MultiSigner>,
+        buffered_single_signature_store: Arc<dyn BufferedSingleSignatureStore>,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            certifier_service,
+            multi_signer,
+            buffered_single_signature_store,
+            logger,
+        }
+    }
+
+    async fn try_register_buffered_signatures_to_current_open_message(
+        &self,
+        signed_entity_type: &SignedEntityType,
+    ) -> StdResult<()> {
+        let discriminant: SignedEntityTypeDiscriminants = signed_entity_type.into();
+        let buffered_signatures = self
+            .buffered_single_signature_store
+            .get_buffered_signatures(discriminant)
+            .await?;
+        let mut signatures_to_remove = vec![];
+
+        for signature in buffered_signatures {
+            match self
+                .certifier_service
+                .register_single_signature(signed_entity_type, &signature)
+                .await
+            {
+                Ok(..) => {
+                    signatures_to_remove.push(signature);
+                }
+                Err(error) => match error.downcast_ref::<CertifierServiceError>() {
+                    Some(CertifierServiceError::InvalidSingleSignature(..)) => {
+                        trace!(self.logger, "Skipping invalid signature for signed entity '{signed_entity_type:?}'";
+                            "party_id" => &signature.party_id,
+                            "error" => ?error,
+                        );
+                    }
+                    _ => {
+                        anyhow::bail!(error);
+                    }
+                },
+            }
+        }
+
+        self.buffered_single_signature_store
+            .remove_buffered_signatures(discriminant, signatures_to_remove)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl CertifierService for BufferedCertifierService {
+    async fn inform_epoch(&self, epoch: Epoch) -> StdResult<()> {
+        self.certifier_service.inform_epoch(epoch).await
+    }
+
+    async fn register_single_signature(
+        &self,
+        signed_entity_type: &SignedEntityType,
+        signature: &SingleSignatures,
+    ) -> StdResult<RegistrationStatus> {
+        match self
+            .certifier_service
+            .register_single_signature(signed_entity_type, signature)
+            .await
+        {
+            Ok(res) => Ok(res),
+            Err(error) => match error.downcast_ref::<CertifierServiceError>() {
+                Some(CertifierServiceError::NotFound(..)) => {
+                    match &signature.signed_message {
```

Comment:
I think that the `SingleSignatures` entity should not be holding a `signed_message`, this is an information that should only be used in the `RegisterSignatureMessage` :thinking: 

In the future we will receive signature messages via other ways than HTTP that will not have a `signed_message` and that we may need to buffer anyway.